src/lib.rs

```diff
@@ -0,0 +1,510 @@
+
+    // The followings are selected restriction lints for rust 1.57
+    clippy::as_conversions,
+    clippy::clone_on_ref_ptr,
+    clippy::create_dir,
+    clippy::dbg_macro,
+    clippy::decimal_literal_representation,
+    // clippy::default_numeric_fallback, too verbose when dealing with numbers
+    clippy::disallowed_script_idents,
+    clippy::else_if_without_else,
+    clippy::exhaustive_enums,
+    clippy::exhaustive_structs,
+    clippy::exit,
+    clippy::expect_used,
+    clippy::filetype_is_file,
+    clippy::float_arithmetic,
+    clippy::float_cmp_const,
+    clippy::get_unwrap,
+    clippy::if_then_some_else_none,
+    // clippy::implicit_return, it's idiomatic Rust code.
+    // clippy::indexing_slicing, TODO
+    // clippy::inline_asm_x86_att_syntax, stick to intel syntax
+    clippy::inline_asm_x86_intel_syntax,
+    clippy::integer_arithmetic,
+    // clippy::integer_division, required in the project
+    clippy::let_underscore_must_use,
+    clippy::lossy_float_literal,
+    clippy::map_err_ignore,
+    clippy::mem_forget,
+    clippy::missing_docs_in_private_items,
+    clippy::missing_enforced_import_renames,
+    clippy::missing_inline_in_public_items,
+    // clippy::mod_module_files, mod.rs file is used
+    clippy::modulo_arithmetic,
+    clippy::multiple_inherent_impl,
+    clippy::panic,
+    // clippy::panic_in_result_fn, not necessary as panic is banned
+    clippy::pattern_type_mismatch,
+    clippy::print_stderr,
+    clippy::print_stdout,
+    clippy::rc_buffer,
+    clippy::rc_mutex,
+    clippy::rest_pat_in_fully_bound_structs,
+    clippy::same_name_method,
+    clippy::self_named_module_files,
+    // clippy::shadow_reuse, it’s a common pattern in Rust code
+    // clippy::shadow_same, it’s a common pattern in Rust code
+    clippy::shadow_unrelated,
+    clippy::str_to_string,
+    clippy::string_add,
+    clippy::string_to_string,
+    clippy::todo,
+    clippy::unimplemented,
+    clippy::unnecessary_self_imports,
+    clippy::unneeded_field_pattern,
+    // clippy::unreachable, allow unreachable panic, which is out of expectation
+    clippy::unwrap_in_result,
+    clippy::unwrap_used,
+    // clippy::use_debug, debug is allow for debug log
+    clippy::verbose_file_reads,
+    clippy::wildcard_enum_match_arm,
+)]
+#![allow(
+    clippy::multiple_crate_versions, // caused by the dependency, can't be fixed
+)]
+
+/// Merge bound trait
+mod merge;
+/// `remove_n` trait
+mod remove_n;
+
+use std::{
+    cmp::Ordering,
+    fmt::Debug,
+    ops::{Bound, RangeBounds},
+};
+
+use clippy_utilities::OverflowArithmetic;
+
+use crate::{merge::BoundMerge, remove_n::RemoveN};
+
+/// Set of Ranges
+#[derive(Debug, PartialEq, Eq, Default, Clone)]
+pub struct MergedRange<K> {
+    /// Sorted ranges
+    ranges: Vec<(Bound<K>, Bound<K>)>,
+}
+
+impl<K> MergedRange<K>
+where
+    K: Ord + Clone,
+{
+    /// New `RangeSet`
+    #[inline]
+    #[must_use]
+    pub fn new() -> Self {
+        Self { ranges: Vec::new() }
```

Comment:
Vec is not an efficient data structure in this case for that ranges are inserted and removed frequently.